pycbc/io/record.py

```diff
@@ -967,4 +1041,353 @@ def add_fields(self, arrays, names=None, assubarray=False):
+        arrays = []
+        names = []
+        for name,dt in fields:
+            arrays.append(default_empty(self.size, dtype=[(name, dt)])) 
+            names.append(name)
+        return self.add_fields(arrays, names)
+
+
+#
+# =============================================================================
+#
+#                           WaveformArray
+#
+# =============================================================================
+#
+
+# we'll vectorize TimeDelayFromEarthCenter for faster processing of end times
+def _time_delay_from_center(tc, detector, ra, dec):
+    return tc + lal.TimeDelayFromEarthCenter(detector.location,
+        ra, dec, tc)
+time_delay_from_center = numpy.vectorize(_time_delay_from_center)
+
+class WaveformArray(_FieldArrayWithDefaults):
+    """A FieldArray with some default fields and properties commonly used
+    by CBC waveforms. This may be initialized in one of 3 ways:
+    
+    1. With just the size of the array. In this case, the returned array will
+    have all of the default field names. Example:
+    >>> warr.fieldnames
+    ('polarization',
+     'mass1',
+     'mass2',
+     'spin2x',
+     'spin2y',
+     'spin2z',
+     'spin1y',
+     'spin1x',
+     'spin1z',
+     'inclination',
+     'coa_phase',
+     'dec',
+     'tc',
+     'ra')
+
+    2. With some subset of the default field names. Example:
+    >>> warr = WaveformArray(10, names=['mass1', 'mass2'])
+    >>> warr.fieldnames
+    ('mass1', 'mass2')
+
+    The list of names may include virtual fields, and methods, as well as
+    functions of these. If one or more virtual fields or methods are specified,
+    the source code is analyzed to pull out whatever underlying fields are
+    needed. Example:
+    >>> warr = WaveformArray(10, names=['mchirp**(5/6)', 'chi_eff', 'cos(coa_phase)'])
+    >>> warr.fieldnames
+    ('spin2z', 'mass1', 'mass2', 'coa_phase', 'spin1z')
+
+    3. By specifying a dtype. In this case, only the provided fields will
+    be used, even if they are not in the defaults. Example:
+    >>> warr = WaveformArray(10, dtype=[('foo', float)])
+    >>> warr.fieldnames
+    ('foo',)    
+
+    Additional fields can also be specified using the additional_fields
+    keyword argument. Example:
+    >>> warr = WaveformArray(10, names=['mass1', 'mass2'], additional_fields=[('lambda1', float)])
+    >>> warr.fieldnames
+    ('mass1', 'mass2', 'lambda1')
+    """
+
+    _staticfields = {
+            'mass1': float,
+            'mass2': float,
+            'spin1x': float,
+            'spin1y': float,
+            'spin1z': float,
+            'spin2x': float,
+            'spin2y': float,
+            'spin2z': float,
+            'tc': float,
+            'coa_phase': float,
+            'ra': float,
+            'dec': float,
+            'polarization': float,
+            'inclination': float,
+    }
+
+    _virtualfields = ['mP', 'mS', 'mtotal', 'q', 'eta', 'mchirp']
+
+    @property
+    def mP(self):
```

Comment:
the names of these methods violate pep8 , please make sure they don't use capitals. https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables